tracing/middleware.go

```diff
@@ -0,0 +1,42 @@
+package tracing
+
+import (
+	"net/http"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+	"github.com/urfave/negroni"
+)
+
+func (t *Tracer) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	var span opentracing.Span
+	opName := r.URL.Path
+
+	// It's very possible that Hydra is fronted by a proxy which could have initiated a trace.
+	// If so, we should attempt to join it.
+	remoteContext, err := opentracing.GlobalTracer().Extract(
+		opentracing.TextMap,
+		opentracing.HTTPHeadersCarrier(r.Header),
+	)
+
+	if err != nil {
+		span = opentracing.StartSpan(opName)
+	} else {
+		span = opentracing.StartSpan(opName, opentracing.ChildOf(remoteContext))
+	}
+
+	defer span.Finish()
+
+	r = r.WithContext(opentracing.ContextWithSpan(r.Context(), span))
+
+	next(rw, r)
+
+	ext.HTTPMethod.Set(span, r.Method)
+	if negroniWriter, ok := rw.(negroni.ResponseWriter); ok {
```

Comment:
Not sure about this line, are you sure this is `negroni`? I think it would be better do define your own interface here or use one from the standard library

Reply:
I am sure that this is the negroni response writer because of https://github.com/urfave/negroni/blob/master/negroni.go#L96

https://github.com/urfave/negroni/blob/master/response_writer.go#L31

With that said, to be extra defensive, I capture [the boolean returned from the type assertion](https://tour.golang.org/methods/15) to verify if the `ResponseWriter` is indeed of type `negroini.ResponseWriter`. All of this is so we can add the tag in our span for the http status that was returned: 

![image](https://user-images.githubusercontent.com/1800781/44953145-42b5fe80-ae45-11e8-986d-398dd79db2e6.png)

Reply:
Ok I see how this works now. This is a negroni middleware so we're getting the correct RW here. I though it worked differently, my bad :)